core-rust/state-manager/src/transaction/executable_logic.rs

```diff
@@ -91,81 +93,111 @@ impl ExecutionConfigurator {
+            "preview".to_string(),
+        )
+    }
+
+    fn wrap_transaction<'a>(
         &'a self,
         executable: Executable<'a>,
         config_type: ConfigType,
+        description: String,
     ) -> ConfiguredExecutable<'a> {
-        ConfiguredExecutable {
+        ConfiguredExecutable::Transaction {
             executable,
             scrypto_interpreter: &self.scrypto_interpreter,
             fee_reserve_config: &self.fee_reserve_config,
             execution_config: self.execution_configs.get(&config_type).unwrap(),
+            threshold: config_type.get_transaction_runtime_warn_threshold(),
+            description,
         }
     }
 }
 
 /// An `Executable` transaction bound to a specific execution configuration.
-pub struct ConfiguredExecutable<'a> {
-    executable: Executable<'a>,
-    scrypto_interpreter: &'a ScryptoVm<DefaultWasmEngine>,
-    fee_reserve_config: &'a FeeReserveConfig,
-    execution_config: &'a ExecutionConfig,
-}
-
-impl<'a> ConfiguredExecutable<'a> {
-    /// Wraps this instance in a time-measuring decorator (which will log a `warn!` after the given
-    /// runtime threshold).
-    pub fn warn_after<D: Display>(
-        self,
+pub enum ConfiguredExecutable<'a> {
+    Flash {
+        flash_receipt: FlashReceipt,
+    },
+    Transaction {
+        executable: Executable<'a>,
+        scrypto_interpreter: &'a ScryptoVm<DefaultWasmEngine>,
+        fee_reserve_config: &'a FeeReserveConfig,
+        execution_config: &'a ExecutionConfig,
         threshold: Duration,
-        description: D,
-    ) -> TimeWarningTransactionLogic<Self, D> {
-        TimeWarningTransactionLogic {
-            underlying: self,
-            threshold,
-            description,
-        }
-    }
+        description: String,
+    },
 }
 
 impl<'a, S: SubstateDatabase> TransactionLogic<S> for ConfiguredExecutable<'a> {
     fn execute_on(self, store: &S) -> TransactionReceipt {
-        execute_transaction(
-            store,
-            self.scrypto_interpreter,
-            self.fee_reserve_config,
-            self.execution_config,
-            &self.executable,
-        )
-    }
-}
-
-/// A time-measuring decorator for a `TransactionLogic`.
-pub struct TimeWarningTransactionLogic<U, D> {
-    underlying: U,
-    threshold: Duration,
-    description: D,
-}
-
-impl<U, D, S> TransactionLogic<S> for TimeWarningTransactionLogic<U, D>
-where
-    S: SubstateDatabase,
-    U: TransactionLogic<S>,
-    D: Display,
-{
-    fn execute_on(self, store: &S) -> TransactionReceipt {
-        let start = Instant::now();
-        let result = self.underlying.execute_on(store);
-        let elapsed = start.elapsed();
-        if elapsed > self.threshold {
-            warn!(
-                "Execution of {} took {}ms, above warning threshold of {}ms",
-                self.description,
-                elapsed.as_millis(),
-                self.threshold.as_millis(),
-            );
+        match self {
```

Comment:
I appreciate this is less composable, but I think that's right actually. I think it is much easier to check/verify correctness having this logic together. And this file is very clearly about "how do I run something against the engine" which is a nice encapsulation.

It was hard to integrate the flash before this refactor because what is 3 lines in the engine is strewn across about 10 files - whereas now we can verify in one place it's configured right :)

The flash is the first transaction on ledger - which basically tells the engine to write substates without executing a transaction - this is to prevent a chicken-and-egg problem with some of the package/system related state that needs to be stored.